skbio/metadata/_testing.py

```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
         self.assertTrue(obj.has_positional_metadata())
+
+
+class IntervalMetadataMixinTests:
+    def _set_up(self):
+        self.upper_bound = 9
+        self.im = IntervalMetadata(self.upper_bound)
+        self.intvls = [
+            {'locations': [(0, 1), (2, 9)], 'metadata': {'gene': 'sagA'}},
+            # repeat it
+            {'locations': [(0, 1)], 'metadata': {'gene': ['a'],
+                                                 'product': 'foo'}}]
+
+    def test_constructor_invalid(self):
+        with self.assertRaisesRegex(TypeError,
+                                    'You must provide `IntervalMetadata` '
+                                    'object.'):
+            self._interval_metadata_constructor_(0, '')
+
+    def test_constructor_interval_metadata_len_mismatch(self):
+        for i in [0, 1, 3, 100]:
+            with self.assertRaisesRegex(ValueError, '\(9\).*\(%d\)' % i):
+                self._interval_metadata_constructor_(i, self.im)
+
+    def test_constructor_no_interval_metadata(self):
+        for i, im in [(0, None), (self.upper_bound, self.im)]:
+            obj = self._interval_metadata_constructor_(i, im)
+            self.assertFalse(obj.has_interval_metadata())
+            self.assertIsInstance(obj.interval_metadata, IntervalMetadata)
+
+    def test_constructor_makes_shallow_copy_of_interval_metadata(self):
+        intvl = self.im.add(**self.intvls[1])
+        obj = self._interval_metadata_constructor_(self.upper_bound, self.im)
+
+        self.assertEqual(obj.interval_metadata, self.im)
+        self.assertIsNot(obj.interval_metadata, self.im)
+
+        # Changing mutable value of metadata of the old interval
+        # also changes obj.
+        intvl.metadata['gene'].append('b')
+        self.assertEqual(obj.interval_metadata, self.im)
+
+        # Changing old interval doesn't change obj
+        intvl.locations = [(3, 6)]
+        self.assertNotEqual(obj.interval_metadata, self.im)
+
+    def test_eq_basic(self):
+        im1 = IntervalMetadata(self.upper_bound)
+        im1.add(**self.intvls[0])
+        obj1 = self._interval_metadata_constructor_(self.upper_bound, im1)
+
+        im2 = IntervalMetadata(self.upper_bound)
+        im2.add(**self.intvls[0])
+        obj2 = self._interval_metadata_constructor_(self.upper_bound, im2)
+
+        self.assertReallyEqual(obj1, obj2)
+
+    def test_eq_populated_differently(self):
+        im1 = IntervalMetadata(self.upper_bound)
+        im1.add(**self.intvls[0])
+        obj1 = self._interval_metadata_constructor_(self.upper_bound, im1)
+
+        obj2 = self._interval_metadata_constructor_(self.upper_bound)
+        obj2.interval_metadata.add(**self.intvls[0])
+
+        self.assertReallyEqual(obj1, obj2)
+
+    def test_eq_both_empty(self):
+        obj1 = self._interval_metadata_constructor_(self.upper_bound)
+        obj2 = self._interval_metadata_constructor_(self.upper_bound)
+        self.assertReallyEqual(obj1, obj2)
+
```

Comment:
This test is important because it asserts `__eq__` doesn't turn an internal `None` representation of interval metadata into an empty `IntervalMetadata` object. The test also makes sense for `metadata` and `positional_metadata`. If the `__eq__` method had been implemented like you had it before (i.e. without the `has_interval_metadata` guards), this test would fail. Basically we don't want to turn `None` into an empty `IntervalMetadata` object for efficiency. If you don't use `has_interval_metadata` to guard that and just use the getter, it will turn `None` into an empty `IntervalMetadata` object.

This has a very real impact on performance when creating many of these objects (e.g. reading a FASTQ file into Sequence objects). We tried removing the `None` optimization and there was a huge degradation in performance. Thus, it is important to test it here and in the other places I requested.